content/blog/2022-05-23-vectorization/index.md

```diff
@@ -0,0 +1,238 @@
+
+The vector language is simulated in the Bril interpreter. I implemented the vector registers as variables storing array of width 4. Each vector operation manipulates the array appropriately; for instance, `veczero` instantiates all the cells of the array to have value 0, `vecload v i d` loads data `d` into the array representing vector `v` at index `i` in the array, and `vecadd v1 v2` does element-wise addition on the arrays representing vectors `v1` and `v2`. The implementation of this interpreter is located in the files `bril-vc` and `brili-vc`. I also augment the `print` instruction to print out all elements in a vector, for easy debugging.
+
+## Vectorization
+
+Vectorization is broken into several stages. The first stage is to preprocess the code, which enables the vectorization algorithms to have greater opportunities to create vectors. The preprocessing stage is comprised of several passes that attempt to create larger basic blocks with more consecutive instructions that can be vectorized, which I call "vectorizable" instructions. 
+
+The second stage is to create runs of vectorizable instructions. A run of instructions is a sequence of Bril instructions that are all have the same opcode and are independent. A vectorizable instruction is independent if its arguments to do not depend on any other destination variables from the same run of instructions. The definitions of independence and same opcode come from Larsen and Amarasinghe's work on superword level parallelism. [^1] For example, the sequence of Bril instructions
+
+```bril
+w: int = add a b;
+x: int = add b c;
+y: int = add a c;
+z: int = add d d;
+```
+
+would be a valid run of vectorizable instructions, while the similar sequence:
+
+```bril
+w: int = add a b;
+x: int = add b w;
+y: int = add a c;
+z: int = add d d;
+```
+
+would not be a valid run of vectorizable instructions because `w` is defined and then used in the same basic block.
+
+Finally, the third stage is to run the vectorization algorithms on the runs of vectorizable instructions, and create vectors, which then replace code in the basic block. Vectors are only created when each run has two or more instructions in it.
+
+### Preprocessing
+
+Preprocessing is applied to the Bril code to allow for further optimization potential. Assuming no memory operations are present, the first preprocessing optimizations applied are simple dead code elimination (DCE) and loop invariant code motion (LICM). Removing obviously dead instructions will help simplify the code. As a side note, the condition for non-present memory conditions is because my implementations do not handle memory operations for DCE or LICM. Also, local value numbering (LVN) would have been used, but my version of LVN had a bug, causing me to exclude using LVN.
```

Comment:
Maybe simplify "non-present memory conditions" to "the absence of memory accesses"?

Reply:
Fixed!